### PR TITLE
Verify sig slasher

### PR DESCRIPTION
**Pull request by shayzluf**
Feature
**What does this PR do? Why is it needed?**
validate signature for is slashable attestation rpc endpoint in slasher
**Which issues(s) does this PR fix?**

Part of #5189

### CODE REVIEW

File: slasher/rpc/server.go [hunk 1]
```diff
@@ -51,3 +69,73 @@ func (ss *Server) IsSlashableAttestation(ctx context.Context, req *ethpb.Indexed
 func (ss *Server) IsSlashableBlock(ctx context.Context, req *ethpb.SignedBeaconBlockHeader) (*slashpb.ProposerSlashingResponse, error) {
 	return nil, errors.New("unimplemented")
 }
+
+func (ss *Server) verifyIndexedAttestationSig(ctx context.Context, indexedAtt *ethpb.IndexedAttestation) (bool, error) {
```
Comment (shayzluf) [inline slasher/rpc/server.go:73]: not sure where should this function live

File: slasher/beaconclient/chain_data.go [hunk 2]
```diff
@@ -27,6 +27,20 @@ func (bs *Service) ChainHead(
...
+// root via gRPC.
+func (bs *Service) GenesisValidatorsRoot(
+	ctx context.Context,
+) ([]byte, error) {
+	ctx, span := trace.StartSpan(ctx, "beaconclient.GenesisValidatorsRoot")
+	defer span.End()
+	res, err := bs.nodeClient.GetGenesis(ctx, &ptypes.Empty{})
+	if err != nil {
+		return nil, errors.Wrap(err, "Could not retrieve genesis data")
+	}
```
Comment (rauljordan) [inline slasher/beaconclient/chain_data.go:40]: Needs a nil pointer check

File: slasher/rpc/server.go [hunk 3]
```diff
@@ -17,17 +25,27 @@ import (
...
 	defer span.End()
-	//TODO(#5189) add signature validation to prevent DOS attack on the endpoint.
+	valid, err := ss.verifyIndexedAttestationSig(ctx, req)
+	if err != nil {
+		log.WithError(err).Error("Failed to verify indexed attestation signature")
+		return nil, status.Errorf(codes.Internal, "Could not verify indexed attestation signature: %v: %v", req, err)
+	}
+	if !valid {
+		log.Warning("Indexed attestation signature failed to verify")
+		return nil, status.Errorf(codes.Unauthenticated, "Indexed attestation signature failed to verify: %v", req)
```
Comment (rauljordan) [inline slasher/rpc/server.go:46]:
```suggestion
		return nil, status.Errorf(codes.Internal, "Indexed attestation signature failed to verify: %v", req)
```

Comment (rauljordan) [reply slasher/rpc/server.go:46]: Unauthenticated is typically used for user auth requests such as not having a proper cookie or auth header, etc.

File: shared/p2putils/fork.go [hunk 4]
```diff
@@ -42,3 +43,36 @@ func CreateForkDigest(
 	}
 	return digest, nil
 }
+
+// Fork given a target epoch and genesis validators root,
+// returns the active fork version in the node.
+func Fork(
```
Comment (rauljordan) [inline shared/p2putils/fork.go:49]: Is this code duplicated also in our p2p package? Maybe we can refactor to just use this shared method if that's the case

File: slasher/rpc/server.go (hunk 1 above)
Comment (rauljordan) [reply slasher/rpc/server.go:73]: Is this duplicated logic from the beacon node? We should likely find some shared package to place it if that's the case

File: shared/p2putils/fork.go (hunk 4 above)
Comment (shayzluf) [reply shared/p2putils/fork.go:49]: i don't see that it is duplicated

File: slasher/rpc/server.go (hunk 3 above)
Comment (shayzluf) [reply slasher/rpc/server.go:46]: thanks, changed

File: slasher/beaconclient/chain_data.go (hunk 2 above)
Comment (shayzluf) [reply slasher/beaconclient/chain_data.go:40]: added

**Comment by shayzluf** [comment]
@rauljordan moved to shared attestationutils

File: slasher/rpc/server.go (hunk 1 above)
Comment (shayzluf) [reply slasher/rpc/server.go]: moved it to attestationsutils

File: beacon-chain/core/blocks/block_operations.go [hunk 5]
```diff
@@ -825,22 +802,8 @@ func VerifyIndexedAttestation(ctx context.Context, beaconState *stateTrie.Beacon
 			pubkeys = append(pubkeys, pk)
 		}
 	}
+	return attestationutil.VerifyIndexedAttestation(ctx, indexedAtt, pubkeys, beaconState.GenesisValidatorRoot(), beaconState.Fork())
```
Comment (nisdas) [inline beacon-chain/core/blocks/block_operations.go:805]: This method now diverges from the spec, it checks the pubkeys first then sorts the attesting indices,etc. Lets keep it the same.

Comment (terencechain) [inline beacon-chain/core/blocks/block_operations.go:805]: This takes in 5 arguments. Can we refactor it to be better?

Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: thanks. chnaged

Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: fixed. thanks

File: beacon-chain/core/blocks/block_operations.go [hunk 6]
```diff
@@ -783,37 +782,17 @@ func ProcessAttestationNoVerify(
 //        return False
 //    return True
 func VerifyIndexedAttestation(ctx context.Context, beaconState *stateTrie.BeaconState, indexedAtt *ethpb.IndexedAttestation) error {
-	ctx, span := trace.StartSpan(ctx, "core.VerifyIndexedAttestation")
+	ctx, span := trace.StartSpan(ctx, "core.VerifyIndexedAttestationSig")
```
Comment (farazdagi) [inline beacon-chain/core/blocks/block_operations.go:785]:
```suggestion
	ctx, span := trace.StartSpan(ctx, "core.VerifyIndexedAttestation")
```
Either the function's name should change to `VerifyIndexedAttestationSig`, or `StartSpan` comment left intact.

Comment (shayzluf) [reply beacon-chain/core/blocks/block_operations.go]: thanks. was auto replaced somehow. good catch

File: shared/attestationutil/attestation_utils.go [hunk 7]
```diff
@@ -68,3 +74,103 @@ func AttestingIndices(bf bitfield.Bitfield, committee []uint64) []uint64 {
 	}
 	return indices
 }
+
+// VerifyIndexedAttestationSig this helper function performs the last part of the
+// spec indexed attestation validation starting at Verify aggregate signature
+// comment.
+//
+// Spec pseudocode definition:
```
Comment (terencechain) [inline shared/attestationutil/attestation_utils.go:82]: Can you update the `pseudocode definition`,  it doesn't reflect to the implementation below

Comment (rauljordan) [reply shared/attestationutil/attestation_utils.go:82]: Updated it to latest spec @terencechain

Comment (shayzluf) [reply shared/attestationutil/attestation_utils.go:82]: i mentioned in the comment above where the implementation starts

**Pull request merged**